TASK-17: Event Page

cristian-aldea opened the pull request:
This PR addressed #70 

This PR aims to implement an event page for use in other user stories.

The images used in the component are placeholders until saving to and querying from the media server is done.

You can access the page by going to http://localhost/test-event-page

Notes:
* Added interfaces for events and related entities, and made the existing entities inherit those interfaces. That way the front-end can use these interfaces.
* Adjusted styling of html, body, and container divs to take up the entire width and height of the page. Without this, the scrolling in the event page info would be much less feasible. 
* Fixed useless test in apps/mull-ui/src/app/components/header/header.spec.tsx
* Moved button from create event page into its own component, under a components/ folder
* created EventRestrictionMap for easy access to enum string values
* Fixed inter font not being loaded due to incorrect url
* although this issue is blocked by #78, I created the PR to get feedback

cristian-aldea committed:
Create event page header component #70

Co-authored-by: Ritchelle <ritchelle.posadas@gmail.com>

cristian-aldea committed:
Add tests for new component #70

Also fixed header test

Co-authored-by: Ritchelle <ritchelle.posadas@gmail.com>

ritchellesap committed:
add event info icons #70

cristian-aldea committed:
Start implementation of event page info component #70

Add interfaces for uses in both the front and back end

Co-authored-by: Ritchelle <ritchelle.posadas@gmail.com>

cristian-aldea committed:
Implement testing for event page info #70

cristian-aldea committed:
Adjust styling of event page info #70

cristian-aldea committed:
Implement mull button #70

cristian-aldea committed:
Refactor Page styling #70

Made html, body and container divs take up 100% of the width and height. This is to make the implementation of scrolling in the event page feasible

cristian-aldea committed:
Implement event page #70

cristian-aldea committed:
Final cleanup before pr #70

cristian-aldea committed:
Remove duplicate css styling #70

cristian-aldea committed:
Update snapshots #70

cristian-aldea committed:
Fix linting errors #70

r-saba reviewed (changes requested):
Looking great! Here's my UI comments

- The green horizontal bar between event info fields is too dark compared to figma
- Missing the last green horizontal bar, there's one under each info row
- The icons on the left are too large
- The event description should have "view more" and once we click that we see the entire text
- Missing the last row, "invite friends"

TheGreatMarkus commented:
Just a note @r-saba, please try to not have too many comments in the main review text, I did that mistake before and it makes it hard to reply (since I can only quote reply)

Here is my response: 
> * The green horizontal bar between event info fields is too dark compared to figma
Sure thing will fix
> * Missing the last green horizontal bar, there's one under each info row
If you look at the mockup you'll see that the divider lines are a bit inconsistent
![image](https://user-images.githubusercontent.com/23544999/99437192-21377500-28e0-11eb-9374-86e34f15e0cb.png)
I tried seeing how it would look if I added lines for the top of the first row and the bottom of the last row, but I didn't like how it looked. We can discuss the specific styling and what we should go for in a meeting
> * The icons on the left are too large
True, ill fix that
> * The event description should have "view more" and once we click that we see the entire text
Ritchelle and I discussed and we felt that that might not be necessary, I can implement it real quick if we need its needed.
> * Missing the last row, "invite friends"
True, didn't include it at the time since a few details we're still shaky, will add it.

r-saba commented:
@TheGreatMarkus Sure will make single comments next time.
Regarding the missing divider at the bottom, do whatever you feel looks nice.
Regarding the "view more", I do think it's necessary. Unless we limit the number of characters in the event description, it could end up taking too much space on this page. If you could implement real quick that'd be sweet thanks!

TheGreatMarkus commented:
@r-saba sounds good, I'll implement the read more thing, and as for the dividers I can ask that real quick during our scrum

cristian-aldea committed:
Address pr comments #70

cristian-aldea committed:
Create expandable text component #70

cristian-aldea committed:
Implement and fix tests #70

TheGreatMarkus commented:
@r-saba implement the expandable text as a separate component, I'm quite happy with how it turned, since its very flexible and easy to use imo. Lemme know what you think, I've addressed all your comments (aside from the div lines, which we'll talk during the meeting today), so I've asked for your review again.

cristian-aldea committed:
Update snapshots #70

RGPosadas commented:
Looked over the UI and I don't have any comments other than good job, it looks and works great! I am just waiting for #78 to be merged so that I can give you a full review.

r-saba commented on apps/mull-ui/src/app/components/event-page-info/event-page-info.scss:
```diff
@@ -0,0 +1,30 @@
+
+.event-page-info-container {
+  div {
+    border-bottom: 1px solid rgba($color: $primary-color, $alpha: 0.5);
```
Matches better the mockup and the mockups use 0.5 for width.
```suggestion
    border-bottom: 0.5px solid rgba($color: $primary-color, $alpha: 0.5);
```

cristian-aldea replied:
Yup, will also simplify this, since we don't need to exclude the last children from having a border

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+
+.title {
+  font-size: 1.5em;
+  font-weight: bold;
```
Title is not bolded in the mockup. 

ps - Since this is on a h1, and we haven't cleared the styles the `font-weight: bold` is unnecessary since it's bold by default styling.

cristian-aldea replied:
You're right, ill make it a div and remove the font weight

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+.title {
+  font-size: 1.5em;
+  font-weight: bold;
+  width: 100%;
```
We're not using flex, by default it will take up the entire space.
```suggestion
```

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+  font-weight: bold;
+  width: 100%;
+  text-align: center;
+  overflow-y: auto;
```
Is there a reason for this?

cristian-aldea replied:
I set a max height for the text of 3 lines (3em) to account for long titles. If the title ends up longer than that I make it scrollable.

cristian-aldea replied:
As we discussed, we will assume title aren't absurdly long

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+  width: 100%;
+  text-align: center;
+  overflow-y: auto;
+  max-height: 3em;
```
Is there a reason for this?

cristian-aldea replied:
See above

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+}
+
+.event-image {
+  max-width: 100%;
```
Typically we use max-width when we want something to stay below 100%. `max-width: 50%`.
This isn't necessary here.
```suggestion
```

cristian-aldea replied:
Removing this

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,49 @@
+.event-image {
+  max-width: 100%;
+  border-radius: 8px;
+  box-sizing: border-box;
```
I don't typically use `border-box` why is this everywhere? When I remove it in the inspector nothing changes

cristian-aldea replied:
Removing this

cristian-aldea replied:
I like to use box-sizing: border-box when I center since otherwise things get misaligned sometimes due to padding. Not necessary in this case.

cristian-aldea replied:
caused by the width:100% in app.scss. Removing all/most of my uses of border-box since its no longer necessary

r-saba commented on apps/mull-ui/src/app/components/event-page-info/event-page-info.scss:
```diff
@@ -0,0 +1,30 @@
+  padding: 0.5rem 1rem;
+  box-sizing: border-box;
+  align-items: flex-start;
+}
```
- `flex-direction` is row by default not need to explicitly define it.
- row will be width 100% by default no need to explicitly define it.
- Do we need `align-items: flex-start`? I don't usually have to use this.

cristian-aldea replied:
Thanks for pointing that out, I'm not familiar with all the default, will be removing these lines

r-saba commented on libs/ui-lib/src/components/mull-button/mull-button.scss:
```diff
@@ -0,0 +1,12 @@
+  background: $primary-color;
+  border-radius: 9px;
+
+  border: none;
```
Button doesn't have a border by default anyway in our app.
```suggestion
.mull-button {
  color: white;
  font-weight: bold;
  padding: 1rem 3rem;
  background: $primary-color;
  border-radius: 9px;
```

cristian-aldea replied:
I think this was left over due to jimmy and jon putting a border for some reason on the original button I moved out of create event.

I probably set it to none just in case, even though its not needed, much like the rest of the dead css in the rest of this pr haha

r-saba commented on libs/ui-lib/src/components/expandable-text/expandable-text.tsx:
```diff
@@ -0,0 +1,41 @@
+            {expand ? 'less' : 'more'}
+          </button>
+        </>
+      )}
```
I would move this into a function and call the function here. Typically we don't want "complex" logic inside the return statement.

cristian-aldea replied:
You would move the tsx here to a function? I don't feel like that's necessary. lets discuss about this together

r-saba commented on libs/ui-lib/src/lib/styles/styles.scss:
```diff
@@ -11,6 +11,13 @@
   border: 0;
 }
 
+html,
+body,
+#root {
+  width: 100%;
+  height: 100%;
+}
+
```
I don't like this, have you considered how it will affect other styling? 
I see you're doing this to keep the button at the bottom I know there was a discussion on if we're keeping it sticky or part of the scrolling section. If it needs to stick to the bottom, consider position abolsute.

cristian-aldea replied:
This isn't to have the button stuck at the bottom. This is to make the event info component scroll correctly doable. For the event info to scroll but not the event page itself, the event page needs to take up the whole page, which wasn't possible since `html`, `body` and `#root` had `fit-content` as their height.

I really doubt this will negatively affect any styling imo, since all it does is make sure that containers take up the entire page, which I thought it was doing already. Also I tested the existing page and no stylng broke as far as I know. 

This warrants a longer conversation, thought, and obviously I might be wrong on what I said earlier. Open to alternatives since you have more experience than me on this.

cristian-aldea replied:
found an alternative solution. The height of the page is 100vh - height of navbar - height of header

r-saba replied:
Nice! That's what I was getting at with `vh` during our discussion, very useful!

r-saba reviewed (changes requested):
Since the PR is strictly front-end I'm paying a lot more attention to the CSS. Just be careful with what you add. CSS that has not been properly thought through can cause a lot of head ache for the next developer when their elements don't style as expected. Also unnecessary/dead CSS is confusing. When in doubt steer towards specificity.

We typically don't have to specify height/width 100% unless we're using something like flex or some other styling is affecting our component.

- Let's replace `em` with `rem` to be consistent. I noticed you were using both in this pr.
- Double check all your CSS

TheGreatMarkus commented:
thanks @r-saba for giving this so much attention, will look into your comments. Just had some quick replies to your comments:

> * Let's replace `em` with `rem` to be consistent. I noticed you were using both in this pr.

We can both use `em` and `rem` when suitable, no needed to force ourselves to use one over the other. I do admit some of my uses of `em` were accidental/bad, but just felt like pointing that out.

> * Double check all your CSS

will do o7

cristian-aldea committed:
Address PR feedback #70

- Fix header title being bold
- remove dead/redundant styling
- simplify border setting in event-info

cristian-aldea committed:
Move components to mull-ui #70

cristian-aldea committed:
Update snapshots #70

r-saba commented on apps/mull-ui/src/app/components/event-page-info/event-page-info.scss:
```diff
@@ -0,0 +1,28 @@
+
+.info-row {
+  display: flex;
+  padding: 0.5rem 1rem;
```
I find the rows could use more breathing room. The spacing in the mockups is not consistent so it's hard to say exactly what we should use. I suggest 1.25rem.
```suggestion
  padding: 1.25rem 1rem;
```

cristian-aldea replied:
Nice catch, I do agree that some more padding is needed, however I feel that 1.25 is excessive from my testing. I bumped up the padding to 1rem:

implementation:
![image](https://user-images.githubusercontent.com/23544999/99762323-3d950680-2ac6-11eb-8770-3b6537ae477b.png)

mockup:
![image](https://user-images.githubusercontent.com/23544999/99762405-6f0dd200-2ac6-11eb-9186-535a316da34b.png)

cristian-aldea committed:
refactor: Improve styling #70

- Remove 100% width from app.scss
- Instead of setting 100% width and height in html, body and #root, use calc to set height of event page to get scrolling to work correctly
- remove useless box-sizing styling
- Moved button to event info

r-saba commented on apps/mull-ui/src/app/components/event-page-info/event-page-info.scss:
```diff
@@ -0,0 +1,28 @@
+
+.event-page-button {
+  align-self: center;
+  margin: 0.5rem 0;
```
Button could use more margin-top to match the mockup and look better.

cristian-aldea replied:
Yup agree, top and bottom should have more margin

r-saba reviewed (changes requested):
Thanks for addressing all my comments. Really good work! I left just a very minor feedback on the UI below.

r-saba commented:
Regarding `em` vs `rem`. I haven't seen the usage of `em` at all actually in web apps. rem is a standard and I suggest we stick with it unless we really have a use case to use `em`. Using both will result in inconsistencies throughout our app.

cristian-aldea committed:
refactor: Move jsx to function #70

cristian-aldea committed:
refactor: Remove history as mull back button prop #70

Instead of passing history throughout out component tree, using the hook is more convenient

cristian-aldea committed:
Add back button to event page #70

cristian-aldea committed:
Update snapshots #70

TheGreatMarkus commented:
> Regarding `em` vs `rem`. I haven't seen the usage of `em` at all actually in web apps. rem is a standard and I suggest we stick with it unless we really have a specific use case to use `em`. Using both will result in inconsistencies throughout our app.

I disagree, from reading articles like https://zellwk.com/blog/rem-vs-em/, each have their usecase. `em` is also a standard, just like `rem`. It's true that so far we haven't had much of a usecase for `em` in our code, but I just don't want the use of `rem` to be dogmatic.

For example, if we end up needing to define padding for different font sizes (p, h1, h2, h3, etc.) that would be a perfect use case for em, since we'd want the padding to change depending on the size of the current element's text.

TheGreatMarkus commented:
Sorry for all the pushes @r-saba, i tried by best to clean up the git history

cristian-aldea committed:
Add more spacing to event page info #70

r-saba commented:
As I mentioned I haven't seen `em` used. Even the website you linked is using rem. Also I've found [plenty ](https://webdesign.tutsplus.com/tutorials/comprehensive-guide-when-to-use-em-vs-rem--cms-23984) of articles suggesting exactly what I said, the usage of rem since it's always consistent and using em only if it is necssary. 

I haven't worked with design systems that spaces content based on font size as you suggest. Typically we have a defined spacing for layout (40px b/w headings and content, 10px between content and buttons etc..). We're saying the same thing that `em` has some use cases in sepcific situations. However we should use `rem` where we can since it's behavior is consistent and easier to predict, and use `em` only if we find a good use case for it. Using both liberally will make our UI look unprofessional.

r-saba commented on apps/mull-ui/src/app/components/mull-back-button/mull-back-button.tsx:
```diff
@@ -1,33 +1,35 @@
       data-testid="mull-back-button"
       className={`mull-back-button ${className}`}
-      onClick={onClick}
+      onClick={onClick ? onClick : defaultOnClick}
```
```suggestion
      onClick={onClick || defaultOnClick}
```

cristian-aldea replied:
I had a feeling there was a better way of doing this. Thanks!

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.scss:
```diff
@@ -0,0 +1,38 @@
+
+.event-datetime {
+  display: flex;
+  flex-direction: row;
```
```suggestion

```

r-saba commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx:
```diff
@@ -0,0 +1,59 @@
+
+export interface EventPageHeaderProps {
+  event: IEvent;
+  lastPage: string;
```
can you rename to `prevPage`? `lastPage` sounds like it's related to pagination

r-saba commented on apps/mull-ui/src/app/components/mull-button/mull-button.scss:
```diff
@@ -0,0 +1,11 @@
+  padding: 1rem 3rem;
+  background: $primary-color;
+  border-radius: 9px;
+
```
Unnecessary spacing
```suggestion
```

r-saba commented on apps/mull-ui/src/app/pages/event-page/event-page.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export interface EventPageProps {
+  event: IEvent;
+  lastPage: string;
```
If you change it to `prevPage` above, change it here too

r-saba reviewed (dismissed):
LGTM! None of my comments below are blockers. If you do address these comments let me know. If you tell me you didn't touch any other part of the code I will approve it quickly after testing functionality.

TheGreatMarkus commented:
@r-saba Yeah we're basically saying the same thing, and I admit my uses of `em` in earlier sprints and in this pr were mostly due to ignorance or error. Just wanted to make sure people reading this understand why we use one over the other.

RGPosadas reviewed (dismissed):
**Subtasks**
- [x] 17.2 Event page header
- [x] 17.3 Event info
- [x] 17.4 Mull button
- [x] 17.5 Integrate Components into this task.

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=2007%3A404529)

**Comments**
Great job! The added padding between the div and the row content (compared to the first time I reviewed this PR) is a great subtle addition.
LGTM!

JimLau49 commented on apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx:
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faCaretRight } from '@fortawesome/free-solid-svg-icons';
+
+import { faClock } from '@fortawesome/free-regular-svg-icons';
+
```
I think we can have svg imports without spaces in between them

cristian-aldea replied:
yup, removed the space

cristian-aldea committed:
Implement pr feedback #70

JimLau49 reviewed (dismissed):
LGTM! , really nice work you have here, i left a small comment as single comment instead of a full review that could be addressed. 

Additionally, ive noticed the caret-right, we're using from FA doesn't exactly match what we have in the mockup.

TASK-17
![image](https://user-images.githubusercontent.com/35787279/99857361-120f2c00-2b59-11eb-8267-5e6b6839cb16.png)

Mockup:
![image](https://user-images.githubusercontent.com/35787279/99857396-26532900-2b59-11eb-8e56-b2927ffa4e51.png)

Im not sure if there is a similar one offered in FA, if not a small edit to the mockup should be done at some point.

Great work 👍

TheGreatMarkus commented:
@JimLau49 
Yeah unfortuantely searching for icons I couldn't find one that matched the mockup. It doesn't really matter anyways

thanks, took a lot longer than I thought!

TheGreatMarkus commented:
Alright I addressed all the comments I saw, lemme know if you see anything else.

JimLau49 reviewed (approved):
Review remains as an approve

r-saba reviewed (approved):
LGTM! Great work
